power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/sparse_lu_solver.hpp

```diff
@@ -328,6 +405,210 @@ template <class Tensor, class RHSVector, class XVector> class SparseLUSolver {
+        solve_once(data, block_perm_array, rhs_.value(), x);
+        Idx num_iter{};
+        // iterate until convergence
+        while (backward_error > epsilon_perturbation) {
+            if (num_iter++ == max_iterative_refinement) {
+                throw SparseMatrixError{};
+            }
+            // calculate residual
+            calculate_residual(x);
+            // solve with residual
+            solve_once(data, block_perm_array, residual_.value(), dx_.value());
+            // calculate backward error and then iterate x
+            backward_error = iterate_and_backward_error(x);
+        }
+        // reset refinement cache
+        reset_refinement_cache();
+    }
+
+    void reset_refinement_cache() {
+        dx_.reset();
+        residual_.reset();
+        rhs_.reset();
+    }
+
+    void initialize_refinement(std::vector<RHSVector> const& rhs, std::vector<XVector>& x) {
+        // save a copy of rhs
+        rhs_ = rhs;
+        // initialize x to zero
+        for (Idx row = 0; row != size_; ++row) {
+            x[row] = XVector{};
+        }
+        // initialize residual to rhs
+        // because r = b - A * x = b - 0 = b
+        residual_ = rhs_;
+        // initialize dx to zero
+        // pre-allocate memory
+        dx_ = x;
+    }
+
+    void calculate_residual(std::vector<XVector> const& x) {
+        auto const& row_indptr = *row_indptr_;
+        auto const& col_indices = *col_indices_;
+        auto const& original_matrix = original_matrix_.value();
+        auto const& rhs = rhs_.value();
+        auto& residual = residual_.value();
+        // calculate residual
+        for (Idx row = 0; row != size_; ++row) {
+            residual[row] = rhs[row];
+            // loop all columns
+            for (Idx idx = row_indptr[row]; idx != row_indptr[row + 1]; ++idx) {
+                // subtract
+                residual[row] -= dot(original_matrix[idx], x[col_indices[idx]]);
+            }
+        }
+    }
+
+    double iterate_and_backward_error(std::vector<XVector>& x) {
+        double backward_error_max_demoniator = 0.0;
+        double backward_error_max_numerator = 0.0;
+        auto const& row_indptr = *row_indptr_;
+        auto const& col_indices = *col_indices_;
+        auto const& original_matrix = original_matrix_.value();
+        auto const& rhs = rhs_.value();
+        auto const& residual = residual_.value();
+        auto const& dx = dx_.value();
+        using RealValueType = std::conditional_t<is_block, Eigen::Array<double, block_size, 1>, double>;
+        // calculate backward error and then iterate x
+        for (Idx row = 0; row != size_; ++row) {
+            // error denominator by |rhs|
+            RealValueType denominator = cabs(rhs[row]);
+            // then append |A| * |x|
+            for (Idx idx = row_indptr[row]; idx != row_indptr[row + 1]; ++idx) {
+                denominator += dot(cabs(original_matrix[idx]), cabs(x[col_indices[idx]]));
```

Comment:
just to double-check is this indeed correct?

$$
(|A|\cdot |x|)_i := \sum_j |a_{ij}| |x_j|
$$

cause the absolute value of a matrix or a vector aren't technically defined in this paper.

EDIT: I looked up the original reference they cited (https://www3.math.tu-berlin.de/Vorlesungen/SS14/MatricesGraphsPDEs/paper_for_students/SIMAX-ardedu-0610013.pdf) and, indeed, there, they do properly define the symbolic notation $$|A|$$ to be the matrix consisting of entries $$a_{ij}$$, and idem for vectors, so this is correct

Reply:
Other papers did mention the meaning of this, as you said.